### PR TITLE
Create context table and use it in various tables

### PR DESCRIPTION
**Pull request by clemens-msupply**
Currently used in user_permission, program, document and document_registry tables.

Part of #1944

# 👩🏻‍💻 What does this PR do? 
Adds a context table and wires it up to the user_permission, program, document and document_registry tables.

Currently, the program context entry row is created automatically during sync using the `program.id` as the `context.id`. Initially I thought this was a bit hacky, but it would allow us to have the context table without extra configuration. Furthermore, the context table doesn't need to be synced, i.e. the context table is just used internally. However, when configuring the permissions you would have to take the `program.id` as the `context.id`. Happy to discuss...

Encounter and enrolment in this diagram are not yet implemented:

```mermaid
erDiagram
    context {
      string id
      string name
    }

    user_permissions {
        string id
        string store_id
        string permission
        string context_id
    }
    context }o--|| user_permissions : context_id

    program {
        string id
        string name
        string context_id
    }
    context }o--|| program : context_id

    document {
        string id
        string context_id
    }
    context }o--|| document : context_id

    document_registry {
        string id
        string context_id
    }
    context }o--|| document_registry : context_id

    encounter {
        string id
        string document_id
        string program_id
    }
    program }o--|| encounter : program_id
    document ||--|| encounter : document_id

    enrolment {
        string id
        string document_id
        string program_id
    }
    program }o--|| enrolment : program_id
    document ||--|| enrolment : document_id
```

# 🧪 How has/should this change been tested? 
Just that unit/integration tests run. Further tests only make sense after the full issues is resolved.

There is a `1944-context-table` branch in the schema repo though to configure the server.

### CODE REVIEW

File: server/service/src/sync/test/integration/central/document_registry.rs
Comment (andreievg) [inline server/service/src/sync/test/integration/central/document_registry.rs:21]: master_list_json ?

File: server/service/src/sync/translations/document_registry.rs
Comment (andreievg) [inline server/service/src/sync/translations/document_registry.rs:1]: dependencies ?

Also would legacy document registry row change to have context_id ? (vs document_context ?)

**Review by andreievg — COMMENTED** [review]
Thanks, looks pretty good, although I am nost sure if I can approve, would this PR make develop not stable ? I mean does it need whole context table refactor before merging to develop ?

I had a few more thoughts after our conversation:

Noticed that patient context is auto generated with migrations, so we should be able to auto add 'patient' to dynamic_permissions. Btw I was just looking at that logic, and i've made a comment earlier, why do we have capabilities as a hash map, and we have CapabilityTag::ContextType as the key we use for allowed_ctx, how doe capabilities work (I couldnt' figure it out in the DSL logic), where would another CapabilityTag come from ?

Last comment, I was wondering, stepping back and considering the extra generic logic introduced by the context, do you think it's still worth the dynamic nature ? Would it have been simpler and faster to have program_id on user_permission (optional) and optional program_id on document and document registry ? I would assume if we use document for something else, we woudl still need to add logic in code for it, at which point we could add another field for permissions ?

Sorry to raise this again, we can discuss this in person, I just wanted to get some reflection on it (most likely we don't have the time to change things now, but still wanted to get your thoughts on it)

**Comment by clemens-msupply** [comment]
> Thanks, looks pretty good, although I am nost sure if I can approve, would this PR make develop not stable ? I mean does it need whole context table refactor before merging to develop ?

Yes, I only plan to merge this PR once the whole issue is resolved.

> Noticed that patient context is auto generated with migrations, so we should be able to auto add 'patient' to dynamic_permissions. Btw I was just looking at that logic, and i've made a comment earlier, why do we have capabilities as a hash map, and we have CapabilityTag::ContextType as the key we use for allowed_ctx, how doe capabilities work (I couldnt' figure it out in the DSL logic), where would another CapabilityTag come from ?

Removing CapabilityTag is part of the issue :) There was a complicated and a bit theoretical reason for it (context name conflicts when the context is not an uuid if I remember correctly, would have to look into it again...) Thus I am quite happy that this can/will be removed since I find it hard myself.

> 
> Last comment, I was wondering, stepping back and considering the extra generic logic introduced by the context, do you think it's still worth the dynamic nature ? Would it have been simpler and faster to have program_id on user_permission (optional) and optional program_id on document and document registry ? I would assume if we use document for something else, we woudl still need to add logic in code for it, at which point we could add another field for permissions ?
> 

I was reflecting on this myself while doing this PR and if the current solution had required extra configs to set up the context entries I would have said lets just go with the program_id solution. Personally I would have done it like in this PR just without the context table, i.e. have the context values be managed in the service layer.  Still feels wrong to me to have hard ties between tables that could be used independently (independent of the question if they ever will be used independently)... Happy to have another chat about it!

File: server/graphql/programs/src/mutations/program_enrolment/insert.rs
Comment (clemens-msupply) [reply server/graphql/programs/src/mutations/program_enrolment/insert.rs:102]: Yes would have to do more renames...

Could catch with the patient id, will fix it here!

File: server/graphql/programs/src/mutations/program_enrolment/update.rs
Comment (clemens-msupply) [reply server/graphql/programs/src/mutations/program_enrolment/update.rs]: Fixed thanks

File: server/repository/src/migrations/v1_01_16/context_table.rs
Comment (clemens-msupply) [reply server/repository/src/migrations/v1_01_16/context_table.rs:27]: thanks will remove the UPDATE lines

File: server/service/src/sync/translations/document_registry.rs
Comment (clemens-msupply) [reply server/service/src/sync/translations/document_registry.rs:1]: Thanks fixes the dependencies.

Yes was thinking of doing that, i.e. having a 4d update once this finished

File: server/repository/src/db_diesel/program_requisition/program_row.rs
Comment (clemens-msupply) [reply server/repository/src/db_diesel/program_requisition/program_row.rs:28]: yes better to have it in the migration, I add it to the issue

File: server/repository/src/db_diesel/document_registry.rs
Comment (andreievg) [reply server/repository/src/db_diesel/document_registry.rs]: category is great !

**Review by andreievg — APPROVED** [review]
Thank you for changes, and thanks for your thoughtful reply

> Still feels wrong to me to have hard ties between tables that could be used independently (independent of the question if they ever will be used independently)... Happy to have another chat about it!

Ooh that's interesting and nicely written. I find it easier to reason about functionality based on data types and their dependencies (especially functionality that I am personally not familiar with), maybe for me it's like strongly typed system and defined schema vs schemaless and dynamic types (the later is much faster to dev with at the start and the former benefits only appear when project grows to a certain size). The reason I was pushing for the context table or program_id was to have descriptive data types with known purpose (and thus needing to know dependencies), so that when the project grows and use case for the data types grow it's very clear how these data types are used and how they shape functionality (kind of like a blueprint), an alternative is to have very generic data types and very thorough documentation for each use case of generic data type (or I guess look at code to see how things work).

Btw above comment does not apply to data types that need to be configurable by user (json forms, plugins etc), only for things that we know before run time and that are common across general use cases.

File: server/repository/src/db_diesel/document_registry.rs
Comment (clemens-msupply) [reply server/repository/src/db_diesel/document_registry.rs]: I will create another PR for the rename...

**Comment by clemens-msupply** [comment]
> Thank you for changes, and thanks for your thoughtful reply
> 
> > Still feels wrong to me to have hard ties between tables that could be used independently (independent of the question if they ever will be used independently)... Happy to have another chat about it!
> 
> Ooh that's interesting and nicely written. I find it easier to reason about functionality based on data types and their dependencies (especially functionality that I am personally not familiar with), maybe for me it's like strongly typed system and defined schema vs schemaless and dynamic types (the later is much faster to dev with at the start and the former benefits only appear when project grows to a certain size). The reason I was pushing for the context table or program_id was to have descriptive data types with known purpose (and thus needing to know dependencies), so that when the project grows and use case for the data types grow it's very clear how these data types are used and how they shape functionality (kind of like a blueprint), an alternative is to have very generic data types and very thorough documentation for each use case of generic data type (or I guess look at code to see how things work).
> 

It depends a bit on the use case on how much schema you want in the database. There are certainly good usecases to have DB constraints in the DB, e.g. invoice and invoice_lines. In this specific case the database integrity is IMHO less important, and you could instead encode some more type saftey in a higher level, e.g. use a Rust `Context` type instead of String ID.

In general, I think its good to have independent components and concepts that you compose and reuse in different ways (only where possible and where it makes sense!). For example, combining tables in the service layer, reusing services in Graphql/Rest layer or having powerful APIs for the FE. When implementing a new feature, you can then go from top to bottom and only make changes where needed. Thus, changes are needed in fewer layers and less code.

Even with most of the business logic encoded in the DB schema you still need docs to describe the higher level concepts. (not sure if a DB schema with lots of cross-references would help me much to understand how things work)

**Pull request merged**
